matchcode/models.py

```diff
@@ -307,22 +314,88 @@ def match(cls, fingerprint, exact_match=False):
+                (hamming_distance, matches)
+            )
 
         if TRACE:
-            for match in good_matches:
-                dct = model_to_dict(match)
-                logger_debug(cls.__name__, 'match:', 'good_matched_package:', dct)
+            for hamming_distance, matches in hamming_distances_and_matches:
+                for match in matches:
+                    dct = model_to_dict(match)
+                    logger_debug(
+                        cls.__name__,
+                        'match:',
+                        'step_3_hamming_distance:',
+                        hamming_distance,
+                        'step_3_matched_package:',
+                        dct
+                    )
 
-        return good_matches
+        # Step 4: use file heuristics to rank matches from step 3
+
+        # If we are not given resource data, return the matches we have
+        if not (resource and hamming_distances_and_matches):
+            remaining_matches = cls.objects.none()
+            if hamming_distances_and_matches:
+                for _, matches in hamming_distances_and_matches:
+                    remaining_matches |= matches
+            return remaining_matches
+
+        resource_size = resource.size
+        matches_by_rank_attributes = defaultdict(list)
+        for hamming_distance, matches in hamming_distances_and_matches:
+            for match in matches:
+                matched_resource = match.package.resources.get(path=match.path)
+
+                if TRACE:
+                    logger_debug(
+                        cls.__name__,
+                        'match:',
+                        'step_4_matched_resource:',
+                        matched_resource
+                    )
+
+                # Compute size and name difference
+                if matched_resource.is_file:
+                    size_difference = abs(resource_size - matched_resource.size)
+                else:
+                    # TODO: index number of files in a directory so we can use
+                    # that for size comparison. For now, we are going to
+                    # disregard size as a factor.
+                    size_difference = 0
+                name_sequence_matcher = SequenceMatcher(a=resource.name, b=matched_resource.name)
+                name_difference = 1 - name_sequence_matcher.ratio()
+                rank_attributes = (hamming_distance, size_difference, name_difference)
+                matches_by_rank_attributes[rank_attributes].append(match)
+
+                if TRACE:
+                    logger_debug(
+                        cls.__name__,
+                        'match:',
+                        'step_4_size_difference:',
+                        size_difference,
+                        'step_4_name_difference:',
+                        name_difference
+                    )
+
+        # Order these from low to high (low being low difference/very similar))
```

Comment:
We should be more explicit about the order of ranking.

```suggestion
        # Order these from low to high (low being low difference/very similar)), first by hamming distance, then by size difference, and finally by name difference.
```